### PR TITLE
better snapshot and fix markdown in notification plain text

### PR DESCRIPTION
**Pull request by Links2004**
see #12, #7

### CODE REVIEW

**Comment by Links2004** [comment]
I have mutlicam support ready if this is merged, since the code is based on the changes here.

https://github.com/Links2004/octoprint-matrix-notifier-plugin/tree/multicam

**Comment by Cadair** [comment]
Thanks a lot for the contribution!

The fact that there isn't a way to get the image with the correct metadata from upstream seems pretty wild.

File: octoprint_matrix_notifier/matrix.py [hunk 1]
```diff
@@ -78,10 +79,11 @@ def upload_media(self, media_data, content_type):
         )
 
     def room_send_markdown_message(self, room_id, text):
+        html = markdown.markdown(text, extensions=['nl2br'])
         content = {
             "msgtype": "m.text",
-            "body": text,
+            "body": BeautifulSoup(html, 'html.parser').get_text(),
```
Comment (Cadair) [inline octoprint_matrix_notifier/matrix.py:85]: What's the objective here? It's reasonably common in the matrix ecosystem to put markdown in the plain text body no?

File: octoprint_matrix_notifier/plugin.py [hunk 2]
```diff
@@ -4,10 +4,13 @@
 from pathlib import Path
 from textwrap import dedent
 
+import requests
+from PIL import Image
+import io
```
Comment (Cadair) [inline octoprint_matrix_notifier/plugin.py:9]: This is already imported.
```suggestion
```

File: octoprint_matrix_notifier/plugin.py [hunk 3]
```diff
@@ -286,20 +271,67 @@ def send_snapshot(self):
...
+
+    def getProxies(self):
+        http_proxy = self._settings.get(["http_proxy"])
+        https_proxy = self._settings.get(["https_proxy"])
+        return {"http": http_proxy, "https": https_proxy}
+
+    def take_image(self, snapshot_url=""):
+        if snapshot_url == "":
+            snapshot_url = self._settings.global_get(["webcam", "snapshot"])
+        if snapshot_url == "":
```
Comment (Cadair) [inline octoprint_matrix_notifier/plugin.py:297]: I think this is a little cleaner:

```suggestion
    def take_image(self, snapshot_url=None:
        if not snapshot_url:
            snapshot_url = self._settings.global_get(["webcam", "snapshot"])
        if not snapshot_url:
```

File: octoprint_matrix_notifier/plugin.py [hunk 4]
```diff
@@ -286,20 +271,67 @@ def send_snapshot(self):
...
             "msgtype": "m.image",
-            "body": file_path.name,
+            "body": time.strftime("%Y_%m_%d-%H_%M_%S") + ".jpg",
             "info": {"mimetype": "image/jpg", "w": img_w, "h": img_h},
             "url": mxc_url,
         }
 
         return self.client.room_send(self.room_id, "m.room.message", content)
+
+    def getProxies(self):
```
Comment (Cadair) [inline octoprint_matrix_notifier/plugin.py:289]:
```suggestion
    @property
    def http_proxy(self):
```

File: octoprint_matrix_notifier/plugin.py [hunk 5]
```diff
@@ -286,20 +271,67 @@ def send_snapshot(self):
...
+                "Please configure the webcam snapshot settings "
+                "before enabling sending snapshots!"
+            )
+            return None
+
+        self._logger.debug("Snapshot URL: " + str(snapshot_url))
+        data = None
+        if snapshot_url:
+            try:
+                r = requests.get(snapshot_url, timeout=10, proxies=self.getProxies())
```
Comment (Cadair) [inline octoprint_matrix_notifier/plugin.py:308]:
```suggestion
                r = requests.get(snapshot_url, timeout=10, proxies=self.http_proxy)
```

File: octoprint_matrix_notifier/plugin.py [hunk 6]
```diff
@@ -286,20 +271,67 @@ def send_snapshot(self):
...
+        if snapshot_url == "":
+            snapshot_url = self._settings.global_get(["webcam", "snapshot"])
+        if snapshot_url == "":
+            self._logger.info(
+                "Please configure the webcam snapshot settings "
+                "before enabling sending snapshots!"
+            )
+            return None
+
+        self._logger.debug("Snapshot URL: " + str(snapshot_url))
```
Comment (Cadair) [inline octoprint_matrix_notifier/plugin.py:304]:
```suggestion
        self._logger.debug("Snapshot URL: %s", snapshot_url)
```

File: octoprint_matrix_notifier/plugin.py [hunk 7]
```diff
@@ -286,20 +271,67 @@ def send_snapshot(self):
...
+        data = None
+        if snapshot_url:
+            try:
+                r = requests.get(snapshot_url, timeout=10, proxies=self.getProxies())
+                data = r.content
+            except Exception as e:
+                self._logger.exception("TimeOut Exception: " + str(e))
```
Comment (Cadair) [inline octoprint_matrix_notifier/plugin.py:311]:
```suggestion
                self._logger.exception("Exception while retrieving snapshot URL: %s", e)
```

**Review by Cadair — COMMENTED** [review]
Just a few little code cleanups, but otherwise looks great thanks.

File: octoprint_matrix_notifier/matrix.py (hunk 1 above)
Comment (Links2004) [reply octoprint_matrix_notifier/matrix.py:85]: the plain text is used for notification in the APP and desktop.
and the notification can not render Markdown.
I personalty can read it fin, but if the notification get pushed small devices like a smart watch then most of the space is lost to Markdown format.

File: octoprint_matrix_notifier/plugin.py (hunk 3 above)
Comment (Links2004) [reply octoprint_matrix_notifier/plugin.py]: update the branch with the multicam support changes which refactor this place ;)

**Pull request merged**